
```{r setup, include=FALSE, results="hide"}
library(adegenet)
library(bit64)
library(cowplot)
library(data.table)
library(ggmap)
library(ggplot2)
library(magick)
library(pander)
library(scales)
library(vcfR)

set.seed(42)

# enable Lato on Ubuntu
my_sis <- Sys.info()["sysname"]
if (my_sis == "Linux") {
    sysfonts::font_add(
        "Lato",
        regular = "/usr/share/fonts/truetype/lato/Lato-Regular.ttf",
        bold = "/usr/share/fonts/truetype/lato/Lato-Bold.ttf",
        italic = "/usr/share/fonts/truetype/lato/Lato-Italic.ttf",
        bolditalic = "/usr/share/fonts/truetype/lato/Lato-BoldItalic.ttf")
}
if(my_sis == "Darwin") {
    sysfonts::font_add(
        "Lato",
        regular = "/Users/tom/Library/Fonts/Lato-Regular.ttf",
        bold = "/Users/tom/Library/Fonts/Lato-Bold.ttf",
        italic = "/Users/tom/Library/Fonts/Lato-Italic.ttf",
        bolditalic = "/Users/tom/Library/Fonts/Lato-BoldItalic.ttf")
}

# knitr options that work with fig_height: 6, fig_width: 8, and fig_crop: false
# in beamer presentations
fh <- grid::convertUnit(unit(227, "pt"), "in", valueOnly = TRUE)
fw <- grid::convertUnit(unit(398, "pt"), "in", valueOnly = TRUE)
knitr::opts_chunk$set(
    dev.args=list(bg = "transparent",
                  family = "Lato",
                  pointsize = 8),
    dev="cairo_pdf",
    echo = FALSE,
    message = FALSE,
    warning = FALSE,
    cache = TRUE,
    fig.align = 'center',
    # out.width = "0.8\\textwidth",
    # out.height  ="0.8\\textheight")
    fig.height = fh * 0.9,
    fig.width = fw,
    out.width = NULL,
    out.height = NULL)
```

```{r globals, include=FALSE, results="hide"}
# colour scheme for plots
viridis_d <- viridis::viridis_pal()
heatscale <- RColorBrewer::brewer.pal(6, "YlOrRd")

pop_order <- c(
    "Coromandel" = "Coromandel",
    "Ruakura" = "Ruakura",
    "Taranaki" = "Taranaki",
    "Wellington" = "Wellington",
    "Greymouth" = "Greymouth",
    "Lincoln" = "Lincoln",
    "O" = "Ophir",
    "Mararoa" = "Mararoa Downs",
    "Mossburn" = "Mossburn",
    "Fortrose" = "Fortrose")

vcf_file <- "data/populations.geo.pruned.vcf"
fai_file <- "data/ref.fasta.fai"
```

```{r data_proc, include=FALSE, results="hide"}
vcf <- read.vcfR(vcf_file)
snp_data <- vcfR2genlight(vcf)

# impute NAs on mean
na_means <- tab(snp_data, NA.method = "mean")
snps_imputed <- new("genlight", na_means)
ploidy(snps_imputed) <- 2

# add pop
pops <- gsub("^[[:alpha:]]+_([[:alpha:]]+).*", "\\1", snps_imputed$ind.names)
pop(snps_imputed) <- factor(pops, levels = names(pop_order))

# set up a plot title
gt <- paste0(length(unique(snps_imputed$ind.names)),
             " individuals genotyped at ", 
             format(snps_imputed$n.loc, big.mark = ","),
             " SNPs")

xv <- xvalDapc(tab(snps_imputed),
               pop(snps_imputed),
               n.pca.max = 20,
               training.set = 0.9,
               result = "groupMean",
               center = TRUE,
               scale = FALSE, 
               n.pca = 1:20,
               n.rep  = 5,
               xval.plot = TRUE)

xv_npca <- which.min(xv$`Root Mean Squared Error by Number of PCs of PCA`)

# run the dapc
dapc_results <- dapc(snps_imputed,
                     n.pca = xv_npca,
                     n.da = length(pop_order) - 1)

dapc_var <- 100 * (dapc_results$eig^2)/(sum(dapc_results$eig^2))

# fst data
summary_file <- "data/populations.sumstats_summary.tsv" 
hetran <- fread(summary_file)[, range(Exp_Het)]

fst_file <- "data/populations.fst_summary.tsv"

fst_mat_dt <- fread(fst_file)

pairwise_fst <- melt(fst_mat_dt,
                     id.vars = "V1",
                     measure.vars = names(fst_mat_dt)[names(fst_mat_dt) != "V1"],
                     variable.name = "pop2",
                     value.name = "Fst")

# fill missing values
setnames(pairwise_fst, "V1", "pop1")
pairwise_fst[pop1 == "MararoaDowns", pop1 := "Mararoa"]
pairwise_fst[pop2 == "MararoaDowns", pop2 := "Mararoa"]

filled_fst <- rbind(pairwise_fst[!is.na(Fst)],
                    pairwise_fst[!is.na(Fst),
                                 .(pop1 = pop2, pop2 = pop1, Fst = Fst)])

all_pop <- pairwise_fst[, unique(c(pop1, as.character(pop2)))]
fst_pd <- rbind(filled_fst,
                data.table(pop1 = all_pop,pop2=all_pop, Fst = 0))

# get population order
fst_mat <- as.matrix(data.frame(dcast(fst_pd, pop1 ~ pop2), row.names = "pop1"))
hc <- hclust(as.dist(fst_mat), method = "ward.D2")
clust_order <- plyr::revalue(hc$labels[hc$order], pop_order)

fst_pd[, pop1 := factor(plyr::revalue(pop1, pop_order),
                        levels = clust_order)]
fst_pd[, pop2 := factor(plyr::revalue(pop2, pop_order),
                        levels = clust_order)]

# set up label colours
my_cols <- structure(rev(viridisLite::viridis(length(pop_order))),
                     names = pop_order)
lab_cols <- plyr::revalue(clust_order, my_cols)

# draw the plot
fill_cols <- RColorBrewer::brewer.pal(7, "YlOrRd")


# fai
fai <- fread(fai_file)

# number of loci per chr
para_file <- "data/populations.rlpara.pruned.vcf"
para_vcf <- fread(cmd = paste("grep -v '^##'", para_file))

n_per_chrom <- merge(para_vcf[, .(`SNPs per contig` = .N),
                              by = `#CHROM`],
                     fai[, .(`#CHROM` = V1,
                             length = V2)])

n_per_chrom[, `SNP frequency` := `SNPs per contig` / length]
med_freq <- n_per_chrom[, median(`SNP frequency`)]
```

```{r nz_map, include=FALSE, results="hide"}
nz_sm <- readRDS("data/nz_map_sm.Rds")
wgs84_loc <- fread("data/manual_locations_with_wgs84.csv")[location != "Reefton"]

wgs84_loc[, loc_code := paste0(location, " (", code, ")")]
setorder(wgs84_loc, -lat)
wgs84_loc[, loc_code := factor(loc_code, levels = rev(loc_code))]
wgs84_loc[, code_loc := paste0(code, " (", location, ")")]

lon_bump <- 0.40
lat_bump <- 0

nz_map <- ggmap(nz_sm) +
    theme_minimal(base_size = 8) +
    theme(legend.key.size = unit(0.5, "lines")) +
    scale_x_continuous(expand = c(0, 0),
                       limits = c(165, 180)) +
    scale_y_continuous(expand = c(0, 0),
                       limits = c(-47, -35)) +
    xlab("Longitude") + ylab("Latitude") +
    geom_point(mapping = aes(x = lon,
                             y = lat,
                             fill = loc_code),
               colour = "black",
               shape = 21,
               size = 2,
               data = wgs84_loc) +
    geom_text(mapping = aes(x = lon + lon_bump,
                            y = lat - lat_bump,
                            label = toupper(location)),
              colour = "black",
              size = 2,
              hjust = "left",
              vjust = 0.5,
              data = wgs84_loc) +
    # fontface = "bold") +
    geom_text(mapping = aes(x = lon - lon_bump,
                            y = lat - lat_bump,
                            label = n),
              colour = "black",
              size = 2,
              hjust = "right",
              vjust = 0.5,
              data = wgs84_loc) +
    # fontface = "bold") +
    scale_fill_viridis_d(guide = FALSE)

```

# Biological control

![](img/Young_Bufo_marinus.jpg){height=70mm}

\source{Bidgee, \href{https://creativecommons.org/licenses/by/3.0}{CC BY}}

\note{\begin{itemize}
\item When you hear biocontrol, you think cane toad
\item Native to south / central america, around 100 introduced into Australia in 1935 to control cane beetle
\item Now over 200 million in Australia, disastrous impact on introduced ecosystem
\end{itemize}}

# Biological control

![](img/Cotesia9061.8.15.07.c.jpg){height=70mm}

\source{Beatriz Moisset, \href{https://creativecommons.org/licenses/by-sa/3.0}{CC BY-SA}}

\note{\begin{itemize}
\item start here - introduce coevolution
\item relies on co-evolution to maintain effectiveness
\item find a good example of successful biocontrol
\item mention that no examples of evolution of resistance
\end{itemize}}


# Argentine stem weevil, *Listronotus bonariensis*

\includegraphics[width=70mm,angle=90]{img/Listronotus_bonariensis_dorsal.jpg}

\source{Landcare Research New Zealand Ltd., \href{https://creativecommons.org/licenses/by/4.0}{CC BY}}

\note{\begin{itemize}
\item today's example is ...
\end{itemize}}

# Pasture in NZ, damage by weevils

\begincols[c]
\begincol{0.5\textwidth}

![](img/Saxon_Rye_Clover_Plus_1024x1024.jpg){height=70mm}

\endcol
\begincol{0.5\textwidth}

![](img/Paddock28_21.2.17.jpg){height=70mm}

\endcol
\endcols

\note{\begin{itemize}
\item it's interesting from the evo point of view but also agriculture
\item importance of pasture to NZ
\item history of ASW in NZ
\item cost of argentine stem weevil damage
\end{itemize}}

# Can't be controlled by insecticides

![](img/Argentine-stem-weevil-larvae-in-ryegrass-stem2.jpg){height=70mm}

\source{\href{http://agpest.co.nz/?pesttypes=argentine-stem-weevil-larvae}{AgPest}}

\note{\begin{itemize}
\item How ASW damages pasture
\item stem borer, insecticide doesn't have access
\item history of ASW in NZ
\item cost of argentine stem weevil damage
\end{itemize}}

# *Microctonus hyperodae*

![](img/1295012-PPT.jpg){height=70mm}
\source{Mark McNeill, \href{https://creativecommons.org/licenses/by/3.0/us/}{CC-BY-3.0}.}

\note{\begin{itemize}
\item Mention that ryegrass \emph{is} an introduced ecosystem
\item Relative lacks of biota etc.
\item mode of action of Mhyp
\end{itemize}}

# Decline in effectiveness

![](img/tomasetto_f1.jpg){width=0.75\textwidth}

\source{PNAS 114 (15): 3885–3890. \href{https://doi.org/10.1073/pnas.1618416114}{10.1073/pnas.1618416114}.
}


\note{\begin{itemize}
\item Mention that ryegrass \emph{is} an introduced ecosystem
\item Relative lacks of biota etc.
\item stress that this is a genetic phenomenon
\item we wanted to look at population genomics so we started by constructing a genome
\end{itemize}}

# Need for a genome

# Poor candidate for short-read genome

\begincols[c]
\begincol{0.6\textwidth}

```{r kmer_plot, fig.width = fw * 0.6}
hist_before_file <- "data/asw_hist.txt"
hist_after_file <- "data/asw_hist-out.txt"
peak_file <- "data/asw_peaks.txt"

peaks <- fread(paste("grep '^[^#]'", peak_file))

hist_files <- c(Raw = hist_before_file, Normalised = hist_after_file)
hist_data_list <- lapply(hist_files, fread)
combined_data <- rbindlist(hist_data_list, idcol = "type")

# arrange plot
combined_data[, type := factor(type, levels = c("Normalised", "Raw"))]

# hlines
mincov <- peaks[2, V1]
p1 <- peaks[2, V2]
maxcov <- peaks[2, V3]
peak_pd <- combined_data[type == "Raw" & between(`#Depth`, mincov, maxcov)]

# plot title
# gt <- paste0(
#     p1, "× 31-mer coverage. ",
#     "Main peak: ", mincov, "×–", maxcov, "×"
# )

# plot
vd <- viridisLite::viridis(3)
line_col <- vd[[1]]
peak_col <- alpha(vd[[2]], 0.5)
ggplot(combined_data,
       aes(x = `#Depth`,
           y = Unique_Kmers,
           colour = type)) +
    theme(legend.position = c(1, 1),
          legend.key.size = unit(1, "lines"),
          legend.justification = c(1,1),
          legend.background = element_blank(),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    geom_path(alpha = 0.75) +
    geom_ribbon(data = peak_pd,
                mapping = aes(ymin = 0,
                              ymax = Unique_Kmers,
                              x = `#Depth`), colour = NA, fill = peak_col,
                show.legend = FALSE) +
    scale_color_viridis_d(guide = guide_legend(title = NULL),
                          direction = -1) +
    scale_y_continuous(
        trans = "log10",
        labels = trans_format("log10", math_format(10^.x)),
        breaks = trans_breaks("log10", function(x) 10^x)) +
    scale_x_continuous(trans = log_trans(base = 4),
                       breaks = trans_breaks(function(x) log(x, 4),
                                             function(x) 4^x)) +
    xlab("Depth") + ylab("Number of unique 31-mers")
```

\endcol
\begincol{0.4\textwidth}

```{r kha_plot, fig.width = fw * 0.4}

CalculateKhaStats <- function(x) {
    setkey(x, `#Depth`)
    x <- x[!last(x)]
    x[, cum_sum := cumsum(as.numeric(Raw_Count))]
    x[, percent_kmers := 100 * cum_sum / sum(Raw_Count)]
    return(x)
}

kha_data_list <- lapply(hist_data_list, CalculateKhaStats)
kha <- rbindlist(kha_data_list, idcol = "type")


kha_plot <- ggplot(kha, aes(x = `#Depth`, y = percent_kmers, colour = type)) +
    theme(legend.justification = c(1,0),
          legend.position = c(1, 0),
          legend.key.size = unit(1, "lines"),
          legend.background = element_blank(),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    scale_color_viridis_d(guide = guide_legend(title = NULL),
                          direction = -1) +
    xlab("Depth") + ylab("Cumulative read 31-mers (%)") +
    ylim(c(0, 100)) +
    scale_x_continuous(trans = log_trans(base = 4),
                       breaks = trans_breaks(function(x) log(x, 4),
                                             function(x) 4^x)) +
    geom_path(alpha = 0.75)


# find the elbows
kha_diff <- kha[`#Depth` <= 256,
                .(diff_pct = diff(diff(Raw_Count) > 0) != 0,
                  `#Depth` = `#Depth`[c(-1, -2)],
                  percent_kmers = percent_kmers[c(-1, -2)]),
                by = type]

kha_diff[, percent_repeat :=
             paste0("~", round(100 - percent_kmers, 0), "% repeats")]

kha_plot +
    geom_hline(data = kha_diff[diff_pct == TRUE][c(5, 21)],
               mapping = aes(yintercept = percent_kmers,
                             colour = type),
               linetype = 2,
               show.legend = FALSE) +
    geom_text(data = kha_diff[diff_pct == TRUE][c(5, 21)],
              mapping = aes(label = percent_repeat,
                            x = 0),
              hjust = "left",
              vjust = -0.1,
              size = 2,
              show.legend = FALSE)

```
\endcol
\endcols

\note{\begin{itemize}
\item started by attempting assembly from illumina reads
\item green shaded area are the homozygous kmers
\item earlier peak (lower depth but more kmers) are heterozygous kmers
\item more kmers are at half depth than at full depth suggesting a highly heterozygous genome
\item knee in kha plot: large proportion of kmers at higher-than-genome coverage, e.g. repeat regions
\item small organism (can't get a lot of DNA from a single indiv)
\item highly heterozygous (pooling not a great idea)
\item don't breed in the lab (can't inbreed)

\end{itemize}}

# We started a genome project by mistake

![](img/withnail.jpg){height=70mm}

\note{\begin{itemize}
\item embracing the sunk cost fallacy
\end{itemize}}

# \emoji1{} Check out these genomes

![](img/genomes.pdf){height=70mm}

\note{\begin{itemize}
\item Vespula species - illumina + HiC.
Really easy to assemble haploid material, you get chromosome-level assemblies
\item stonefly. PacBio Hifi
\item velvetworm, 12 GB genome, medium coverage nanopore (20x or so?) takes months to assemble, 
\end{itemize}}

# Iterative improvements to the genome

```{r genome_table, results="asis", cache=FALSE}
genome_table <- fread("data/genome_table.csv", header = TRUE)

setnames(genome_table, names(genome_table)[[1]], "rn")
x <- data.frame(genome_table, row.names = "rn")
panderOptions("keep.line.breaks", TRUE)
pandoc.table(x,
             col.names = names(genome_table)[2:length(names(genome_table))],
             # style = "rmarkdown", 
             style = "multiline",
             split.tables = Inf)
```

\source{Insects 11 (7): 441. \href{https://doi.org/10.3390/insects11070441}{10.3390/insects11070441}.
}

\note{\begin{itemize}
\item RNAseq mapping - 75%
\item long reads - 99.88 %
\item short reads - 94.71 %
\end{itemize}}

# Nanopore genome

<!-- ![](img/screenshot-www.ncbi.nlm.nih.gov-2020.09.15-15_27_28.png){height=70mm} -->

\source{Insects 11 (7): 441. \href{https://doi.org/10.3390/insects11070441}{10.3390/insects11070441}.
}

\note{\begin{itemize}
\item The genome is now on NCBI
\end{itemize}}

# Geographical survey

```{r geo_surv}
print(nz_map)
```
\source{
Map tiles by
\href{http://stamen.com}{Stamen Design} under \href{http://creativecommons.org/licenses/by/3.0}{CC BY 3.0}, with data by \href{http://openstreetmap.org}{OpenStreetMap} under \href{http://www.openstreetmap.org/copyright}{ODbL}.}

# `r format(snps_imputed$n.loc, big.mark = ",")` SNPs after processing, calling and filtering

```{r filter_map}

geo_only <- readRDS("data/frac_discarded.geo.Rds")
geo_only[, pop2 := factor(pop, levels = rev(pop_order))]
flagstats <- readRDS("data/flagstats.Rds")
mean_dp <- readRDS("data/mean_dp.Rds")

pd1 <- rbind(geo_only[!is.na(pop2),
                      .(pop2,
                        indiv = individual,
                        value = frac_discarded *100,
                        variable = "Bases discarded (%)")],
             flagstats[!is.na(pop2) & category == "mapped %",
                       .(pop2, indiv, value = qc_pass, variable = "Mapped (%)")])

pd <- rbind(pd1,
            mean_dp[!is.na(pop2), .(
                pop2, indiv, value = mean_dp, variable = "Mean depth across loci"
            )])

ggplot(pd, aes(x = pop2, y = value, col = pop2)) +
    facet_wrap(~variable,
               scales = "free_x",
               strip.position = "bottom") +
    theme(strip.placement = "outside",
          strip.background = element_blank()) +
    coord_flip() +
    xlab(NULL) +
    ylab(NULL) +
    scale_colour_viridis_d(guide = FALSE) +
    geom_boxplot(fill = NA,
                 colour = alpha("black", 0.5),
                 outlier.colour = NA,
                 width = 0.5) +
    geom_point(shape = 16,
               alpha = 0.8,
               position = position_jitter(width = 0.2))
```


\note{
1-1.5 million reads per individual
}

\note{
- need stuff about the adaptor trimming strategy here
- number of loci, coverage per locus
}

# `r gt`

\begincols[c]
\begincol{0.5\textwidth}

```{r popgen1, fig.width = fw * 0.5}
print(nz_map)
```

\endcol
\begincol{0.5\textwidth}

```{r popgen2, fig.width = fw * 0.5}
# run the pca
pca <- glPca(snps_imputed, nf = Inf)
pct_var <- 100 * (pca$eig^2)/(sum(pca$eig^2))

# generate pca plot data
pca_dt <- data.table(pca$scores, keep.rownames = TRUE)
setnames(pca_dt, "rn", "individual")
pca_dt[, population := gsub("^[[:alpha:]]+_([[:alpha:]]+).*", "\\1", individual)]
pca_dt[, population := factor(
    plyr::revalue(as.character(population), pop_order),
    levels = pop_order)]

ggplot(pca_dt, aes(x = PC1, y = PC2, colour = population)) +
    coord_fixed() +
    scale_colour_viridis_d(direction = -1,
                           guide = FALSE) +
    xlab(paste0("PC1 (", round(pct_var[[1]], 1), "%)")) + 
    ylab(paste0("PC2 (", round(pct_var[[2]], 1), "%)")) + 
    geom_point(alpha = 0.8, shape = 16)
```

\endcol
\endcols

# DAPC of PCs 1--`r xv_npca`

```{r popgen4}
# generate dapc plot data
dapc_dt <- data.table(dapc_results$ind.coord, keep.rownames = TRUE)
setnames(dapc_dt, "rn", "individual")
dapc_long <- melt(dapc_dt,
                  id.vars = "individual",
                  variable.name = "component",
                  value.name = "score")
dapc_long[, population := gsub("^[[:alpha:]]+_([[:alpha:]]+).*", "\\1", individual)]
dapc_long[, population := factor(plyr::revalue(as.character(population), pop_order),
                                 levels = pop_order)]

dapc_pd <- merge(dapc_long,
                 data.table(component = paste0("LD", 1:length(dapc_var)),
                            dapc_var = dapc_var),
                 all.y = FALSE)
# dapc_pd[, population := factor(population, levels = pop_order)]
dapc_pd[, facet_label := paste0(component, " (", round(dapc_var, 1), "%)")]

# plot the dapc
ggplot(dapc_pd[dapc_var > 1],
       aes(x = population, y = score, colour = population)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          panel.border = element_rect(fill = NA, colour = "black"),
          strip.placement = "outside",
          strip.background = element_blank()) +
    xlab(NULL) + ylab(NULL) +
    ylim(c(-6, 6)) +
    facet_wrap(~facet_label,
               strip.position = "left",
               scales = "free") +
    scale_colour_viridis_d(direction = -1,
                           guide = FALSE) +
    geom_boxplot(fill = NA,
                 colour = alpha("black", 0.5),
                 outlier.colour = NA,
                 width = 0.5) +
    geom_point(shape = 16,
               alpha = 0.8,
               position = position_jitter(width = 0.2))
```

# Assigned population

```{r dapc_post}

# posteriors
dapc_post_wide <- data.table(dapc_results$posterior, keep.rownames = TRUE)
dapc_post <- melt(dapc_post_wide, id.vars = "rn", variable.name = "pop", value.name = "prob")
dapc_post[, sample_pop := gsub("geo_([[:alpha:]]+).*", "\\1", rn)]

dapc_post[, pop := factor(plyr::revalue(as.character(pop), pop_order),
                          levels = rev(pop_order))]
dapc_post[, sample_pop := factor(plyr::revalue(as.character(sample_pop), pop_order),
                                 levels = pop_order)]

actual_pop <- dapc_post[pop == sample_pop]
setorder(actual_pop, -prob)
indiv_order <- actual_pop[, rn]

dapc_post[, rn := factor(rn, levels = unique(indiv_order))]

dp <- ggplot(dapc_post, aes(x = rn, y = prob, fill = pop)) +
    theme_minimal() +
    theme(
        # text = element_text(debug = TRUE),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 4.5),
        strip.text = element_text(size = 4.5),
        strip.background = element_blank(),
        legend.key.size = unit(0.8, "lines"),
        legend.position = "top",
        panel.background = element_blank(),
        panel.spacing = unit(1/10, "lines")) +
    xlab("Sample population") + ylab("Membership probability") +
    facet_grid(~ sample_pop,
               scales = "free_x",
               space = "free_x",
               switch = "x", ) +
    scale_y_continuous(expand = c(0, 0)) +
    scale_fill_viridis_d(
        direction = 1,
        guide = guide_legend(title = NULL,
                             title.position = "top",
                             reverse = TRUE)) +
    geom_col()

gb <- ggplot_build(dp)
gt <- ggplot_gtable(gb)
gt$layout$clip <- "off"
grid::grid.draw(gt)

```

# High heterozygosity (`r sprintf('%.02f', hetran[[1]])`--`r sprintf('%.02f', hetran[[2]])`), low structure

\begincols[c]
\begincol{0.5\textwidth}

```{r popgen7, fig.width = fw * 0.5}
print(nz_map)
```

\endcol
\begincol{0.5\textwidth}

```{r pairwise_fst, fig.width = fw * 0.5}
ggplot(fst_pd, aes(x = pop1, y = pop2, fill = Fst)) +
    theme_minimal(base_size = 8 ) +
    theme(axis.text.x = element_text(angle = 90,
                                     hjust = 1,
                                     vjust = 0.5,
                                     colour = lab_cols),
          axis.text.y = element_text(colour = lab_cols),
          legend.key.size = unit(0.8, "lines")) +
    xlab(NULL) + ylab(NULL) +
    scale_x_discrete(expand = c(0, 0)) +
    scale_y_discrete(expand = c(0, 0)) +
    # scale_y_discrete(limits = rev(levels(fst_pd$pop2)),
    #                  expand = c(0, 0)) +
    scale_fill_gradientn(
        colours = fill_cols,
        guide = guide_colorbar(
            title = expression(italic("F")["ST"]))) +
    geom_raster()

```

\endcol
\endcols

# Previous measures of genetic diversity

- maybe a RAPD gel? or something random

\note{\begin{itemize}
\item molecular markers go here.
\item "there is a high degreeof genetic similarity between populations throughoutthe country" (RAPD paper)
\item "Only one haplotype of COI was found in all of the L. bonariensis specimens collected from eight sites throughout New Zealand" (COI sequencing)
\end{itemize}}

# Introduce unequal arms race

# Main divide separates ASW populations

\begincols[c]
\begincol{0.5\textwidth}

```{r popgen5, fig.width = fw * 0.5}
print(nz_map)
```

\endcol
\begincol{0.5\textwidth}

```{r popgen6, fig.width = fw * 0.5}
ggplot(dapc_pd[component == "LD1"],
       aes(x = population, y = score, colour = population)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          panel.border = element_rect(fill = NA, colour = "black"),
          strip.placement = "outside",
          strip.background = element_blank()) +
    xlab(NULL) + ylab(NULL) +
    ylim(c(-6, 6)) +
    facet_wrap(~facet_label,
               strip.position = "left",
               scales = "free") +
    scale_colour_viridis_d(direction = -1,
                           guide = FALSE) +
    geom_boxplot(fill = NA,
                 colour = alpha("black", 0.5),
                 outlier.colour = NA,
                 width = 0.5) +
    geom_point(shape = 16,
               alpha = 0.8,
               position = position_jitter(width = 0.2))

```

\endcol
\endcols

# Demographic modelling

```{r fsc}


# model 0: iv, model 1: i, model 2: iii, model 3: ii, model 4: v
model_map <- c(
    model1 =   "i",
    model3 =  "ii",
    model2 = "iii",
    model0 =  "iv",
    model4 =   "v")

# asw-stacks-singleplate/output/095_fastsimcoal/ns.pruned/summary.csv
model_file <- "data/summary.csv"
model_diagram <- image_read_pdf("data/models_fig_170ns.pdf")

# set up labels
model_data <- fread(model_file)
model_data[, plot_model := factor(plyr::revalue(model, model_map),
                                  levels = model_map)]
model_data[mig == "mig", mig := "With migration"]
model_data[mig == "no_mig", mig := "Without migration"]

# deal with inf
max_lhood <- model_data[is.finite(Lhood), max(Lhood)]
model_data[, is_inf := is.infinite(Lhood)]
model_data[is_inf == TRUE, Lhood := max_lhood]

# summary table?
model_summary <- model_data[, data.table(t(summary(Lhood))), by = .(plot_model, mig)]
# model_summary[V2 == "Mean"]

lhood_plot <- ggplot(model_data[is_inf == FALSE],
                     aes(x = plot_model,
                         y = Lhood,
                         colour = mig)) +
    theme_grey(base_size = 10) +
    theme(legend.key.size = unit(0.8, "lines")) +
    scale_color_viridis_d(
        guide = guide_legend(title = NULL,
                             override.aes = list(shape = 16, alpha = 1))) +
    xlab(NULL) + ylab("Delta likelihood") +
    geom_point(position = position_jitterdodge(jitter.width = 0.25),
               alpha = 0.5,
               size = 1,
               shape = 16) +
    geom_point(data = model_data[is_inf == TRUE],
               position = position_jitterdodge(jitter.width = 0.25),
               alpha = 0.5,
               size = 1,
               shape = 8)


plot_grid(ggdraw() + draw_image(model_diagram),
          lhood_plot,
          align = "h",
          axis = "l",
          ncol = 1)

```

\source{Marissa Le Lec}

\note{
\begin{itemize}
\item Why are the models so weird? Look at diagnostic plot.
\end{itemize}
}

# SNPs don't diff para


```{r rlpara_prep, results = "hide"}
para_info_file <- "data/para_sample_info.tsv"


# get the info
para_info <- fread(para_info_file)
# mung to match
para_info[, indiv_name := paste("para", Individual, sep = "_")]

# read the vcf
para <- read.vcfR(para_file)
para_data <- vcfR2genlight(para)

# impute NAs on mean
para_means <- tab(para_data, NA.method = "mean")
para_imputed <- new("genlight", para_means)
ploidy(para_imputed) <- 2

# add pop
p_pops <- para_info[para_imputed$ind.names, pop_para, on = "indiv_name"] 
pop(para_imputed) <- factor(p_pops)

para_pca <- glPca(para_imputed, nf = Inf)
para_pct_var <- 100 * (para_pca$eig^2)/(sum(para_pca$eig^2))

# generate pca plot data
para_pca_dt <- data.table(para_pca$scores, keep.rownames = TRUE)
setnames(para_pca_dt, "rn", "individual")
para_pca_dt[, population := para_info[individual, pop_para, on = "indiv_name"]]
para_pca_dt[, c("loc", "para") := tstrsplit(population, "_")]
paras <- c("non-parasitized" = "Not detected",
           "parasitized" = "Detected")

para_pca_dt[, para := plyr::revalue(para, paras)]
```

```{r rlpara}
ggplot(para_pca_dt, aes(x = PC1,
                        y = PC2,
                        colour = para,
                        shape = loc)) +
    coord_fixed() +
    scale_colour_viridis_d(direction = -1,
                           guide = guide_legend(
                               title = "Parasitoid",
                               override.aes = list(size = 1)
                           )) +
    scale_shape_manual(values = c(16, 17),
                       guide = guide_legend(
                           title = NULL,
                           override.aes = list(size = 1))) +
    xlab(paste0("PC1 (", round(para_pct_var[[1]], 1), "%)")) + 
    ylab(paste0("PC2 (", round(para_pct_var[[2]], 1), "%)")) + 
    geom_point(alpha = 0.8, size = 3)
```

\source{Marissa Le Lec}


# What sort of variation do we expect biocontrol to act on?

\begincols[c]
\begincol{0.5\textwidth}

![](img/selection.png){height=70mm}

\endcol
\begincol{0.5\textwidth}

![](img/F1.large.jpg){height=70mm}

\endcol
\endcols

\source{
Trends in Ecology \& Evolution 9 (5): 166–169.
\href{http://www.sciencedirect.com/science/article/pii/0169534794900795}{10.1016/0169-5347(94)90079-5}; 
PNAS 116 (21): 10424–10429. \href{https://doi.org/10.1073/pnas.1821713116}{10.1073/pnas.1821713116}.
}


\note{
\begin{itemize}
\item Dros / insecticide models of selection
\item biocontrol never selects outside the fitness distribution, so we don't expect large-effect variants
\item anthranilic diamides (Chlorantraniliprole) target the ryanodine receptor which is not like earlier insecticides that target neurotransmission
\item has not been selected for but the genetic architechture for resistance is scattered across the genome in weak variants
\item the massive heterozygosity in ASW populations is consistent with this
\end{itemize}
}


# One SNP every `r format(round(1/med_freq, 0), big.mark = ",")` bp

\begincols[c]
\begincol{0.4\textwidth}

```{r snp_per_contig, fig.width = fw * 0.4}
ggplot(n_per_chrom, aes(x = `SNPs per contig`)) +
    ylab("Contigs") +
    scale_x_continuous(breaks= pretty_breaks()) +
    geom_histogram(binwidth = 1)
```

\endcol
\begincol{0.6\textwidth}

```{r snp_freq, fig.width = fw * 0.6}
ggplot(n_per_chrom[`SNPs per contig` > 1],
       aes(x = `SNP frequency`)) +
    ylab("Contigs") + xlab("Mean SNP frequency") +
    scale_x_continuous(
        trans = "log10",
        labels = trans_format("log10", math_format(10^.x)),
        breaks = trans_breaks("log10", function(x) 10^x)) +
    geom_histogram(bins = 100)
```

\endcol
\endcols

\note{
One every 38 kb is median for the rlpara dataset
}

# Mention RNAseq here?

![](img/sarah_rnaseq.png){height=70mm}

\source{Sarah Inwood}

\note{\begin{itemize}
\item introduce the wasp biology here
\item what are the mechanisms by which ASW has escaped control?
\item talk about immune suppression (manduca example)
\item immunity? behaviour?
\item maybe compare to \emph{M. aethiopoides}
\end{itemize}}

# Parasitoid genomes

![](data/wga.pdf)

\source{John Skelly}

# Genetic diversity in the parasitoid?

\source{Ciar\'{a}n Cuddy}

\note{\begin{itemize}
\item note implications for monitoring biocontrol
\item maybe compare to \emph{M. aethiopoides}
\end{itemize}}


# Ppl to thank

- Goldson, Malvika, Lokesh
- Weevil dissectors
- 


